Backup process. (#1121)
btcpayserver/btcpayserver-doc: docs/Deployment/BackupRestore.md

# Backup & Restore

This guide gets you up to speed with the [Docker deployment](../Docker/)'s Backup & Restore process.
You will learn about what to keep in mind when doing a backup and how to restore a backup.

[[toc]]

## Remarks and Considerations

> The original backups strategy in BTCPay Server still exists and can be found [here](https://docs.btcpayserver.org/Docker/#how-can-i-back-up-my-btcpay-server).

Reviewer: this page need to have a BIG warning pointing out to this new process. (or maybe just delete the old process?)

Reviewer 2: Updated that part in btcpayserver/btcpayserver-docker#655.

With these preparations taken, the backup process is now starting. 

The script has checks to ensure it either works or fails with a comprehensive error message at every step of the way.
This gets determined by a `-e` tag at the start of the script. 

```
echo "🚨 Database container could not be started or found."
```

If everything is running smoothly, you'll get to see multiple completed marks in your console. 
Whenever the backup has been completed successfully, it will state:

```
printf "✅ Backup done => $backup_path\n\n"
```

Your BTCPay Server has now finished the backup process. It's now up to you to store these backups in a safe manner. 
After you've made a backup the first time, it's always wise to at least test your backup in a restore scenario. 
In the next topic, we will go over the extra options you can set with your backup.

## Extra options for `btcpay-backup.sh`

When you run the `btcpay-backup.sh` script, you're able to set certain flags.

### Set a Passphrase for BTCPay Server backup. 

One of the extra tags that you can set when running the `btcpay-backup.sh` is for a passphrase.
This would be run as follows :

```
cd "$BTCPAY_BASE_DIRECTORY"

> ./btcpay-backup.sh BTCPAY_BACKUP_PASSPHRASE 

Reviewer 3: you just put the passphrase behind the backup script? Without any option flag introducing it? Seems a bit weired for me. Also the passphrase would now exist in the shell(bash)'s history file. Might be worth mentioning?

Reviewer 2: I've now mentioned it in the text, but I think we should solve it using another way, like using an env file, so that it doesn't end up in the history.

Creating generated_postgres_1    ... done
Creating generated_bitcoin_rtl_1 ... done
Creating tor                     ... done
Creating btcpayserver_bitcoind   ... done
Creating tor-gen                         ... done
Creating generated_nbxplorer_1           ... done
Creating btcpayserver_clightning_bitcoin ... done
Creating generated_clightning_bitcoin_charge_1 ... done
Creating generated_clightning_bitcoin_spark_1  ... done
Creating generated_btcpayserver_1              ... done
Creating generated_clightning_bitcoin_rest_1   ... done
Creating nginx-gen                             ... done
Creating letsencrypt-nginx-proxy-companion     ... done

ℹ️  Cleaning up …

✅ Backup done => /var/lib/docker/volumes/backup_datadir/_data/backup.tar.gz.gpg

```

## How does the BTCpay-restore.sh work? 

It's very similar to the `btcpay-backup.sh` process but in reverse. 
The `btcpay-restore.sh` script needs to be run with the path to your `backup.tar.gz` file. 

First off, open a terminal and type the following as root.  
Remember that if you set `BTCPAY_BACKUP_PASSPHRASE` on the backup, you also need to provide it for decryption :

```
cd "$BTCPAY_BASE_DIRECTORY"

> export BTCPAY_BACKUP_PASSPHRASE="tOpSeCrEt"

Reviewer 3: I find it a bit inconsitent that for backup you specifiy the passphrase as an option and for restore you have to set an env variable. Couldn't we make the workflow a bit more consitant? Question towards @dennisreimann

Reviewer 2: It works via env variable, I've updated the instructions.

./btcpay-restore.sh /var/backups/backup.tar.gz.gpg
```

This will start the restore process by unpacking the backup file.

> If it can't find the file in the provided path, the script will give out an error.  

Reviewer 3: ```suggestion
If it can't find the file in the provided path, the script will exit with an error.  
```

volumes/generated_btcpay_datadir/_data/btcpay20220512.log
volumes/generated_btcpay_datadir/_data/tmp/
volumes/generated_btcpay_datadir/_data/host_id_rsa
volumes/generated_btcpay_datadir/_data/Main/
volumes/generated_btcpay_datadir/_data/Main/settings.config
volumes/generated_btcpay_datadir/_data/host_id_rsa.pub
volumes/generated_btcpay_datadir/_data/host_authorized_keys
volumes/generated_btcpay_datadir/_data/key-47ec2b82-ed01-45ef-8a63-390470cc10e5.xml
[...]
volumes/generated_tor_torrcdir/
volumes/generated_tor_torrcdir/_data/
volumes/generated_tor_torrcdir/_data/torrc
volumes/generated_tor_torrcdir/_data/torrc-2

ℹ️  Restoring database …--
-- PostgreSQL database cluster dump
--
SET default_transaction_read_only = off;
SET
SET client_encoding = 'UTF8';
SET
SET standard_conforming_strings = on;
SET
--
-- Drop databases (except Postgres and template1)
--
DROP DATABASE btcpayservermainnet;
DROP DATABASE nbxplorermainnet;
ERROR:  database "btcpayservermainnet" is being accessed by other users
DETAIL:  There is 1 other session using the database.

[...]

--
-- PostgreSQL database dump complete
--
--
-- PostgreSQL database cluster dump complete
--

ℹ️  Cleaning up …

✅ Restore done
```

> ## What isn't included and should be considered before taking this backup approach. 

Reviewer 3: Maybe introduce a seperate FAQ heading for these "question" type headings?

Reviewer 2: Updated that part.

### Lightning channel backup. 

:::warning
It has a high chance of failure in a disaster recovery scenario, where you may do a backup once per night and need to restore that one backup.

> For instance, the Lightning static channel backup should be watched by a script and copied over to a remote server to ensure you always have the latest state available. The scripts will be sufficient in a migration case, where the shutdown of the old and the start of the new server happen directly. 

Reviewer 3: I would very much appreciated a Link to the SCB solution of LND at this point.

Reviewer 2: We will add this as a separate backup script. This is to be done.

Author: If you do happen to more things to consider when running the backup LMK. 
I think it's wise, just as I stated when you start reading this document, that BTCPay in it self can't and won't be responsible for it in anyway. The more we can make the users aware of the better and less questions will be raised.

# Back up & Restore. 

In this guide, you will get up to speed with BTCPay Server's `Backup & Restore` process, what you need to keep in mind when doing a backup, what's included, and how to restore a backup. 

:::warning
BTCPay Server is and will never be responsible for your backup. 
This process comes with its warnings, and they should be considered. 
Always make sure your backup solution works and fits your needs! 
:::

## How does BTCPay Server backup work? 

> The original strategy of backups in BTCPay Server still exists and can be found [Here](https://docs.btcpayserver.org/Docker/#how-can-i-back-up-my-btcpay-server).
> While we did create a new process, the old scripts still do work and have only slightly been touched. see [Github #641](https://github.com/btcpayserver/btcpayserver-docker/pull/641) for more details. 

Reviewer 2: ```suggestion
The original strategy of backups in BTCPay Server still exists and can be found [here](https://docs.btcpayserver.org/Docker/#how-can-i-back-up-my-btcpay-server).
While we did create a new process, the old scripts still do work and have only slightly been touched. See [pull request #641](https://github.com/btcpayserver/btcpayserver-docker/pull/641) for more details. 
```

echo "✅ Archive done."
```

> When the backup has been completed and run successful it will state :

Reviewer 2: ```suggestion
When the backup has been completed successfully it will state:
```

```

## How does the BTCpay-restore.sh work? 

It's very similar to the `btcpay-backup.sh` process but in reverse. 

> The `btcpay-restore.sh` script will search for the `backup.tar.gz` file. 
> 
> First off, open a terminal and type the following as root. Remember that if you set `BTCPAY_BACKUP_PASSPHRASE` on the backup, you have to set it on the restore too ; 

Reviewer 2: ```suggestion
The `btcpay-restore.sh` script needs to be run with the path to your `backup.tar.gz` file. 

First off, open a terminal and type the following as root. Remember that if you set `BTCPAY_BACKUP_PASSPHRASE` on the backup, you have to set it on the restore too. 
```

```
cd "$BTCPAY_BASE_DIRECTORY"

> ./btcpay-restore.sh BTCPAY_BACKUP_PASSPHRASE 

Reviewer 2: ```suggestion
./btcpay-restore.sh /var/backups/backup.tar.gz

# if you used the passphrase for encryption, you also need to provide it for decryption
export BTCPAY_BACKUP_PASSPHRASE="tOpSeCrEt"
./btcpay-restore.sh /var/backups/backup.tar.gz.gpg
```

```

> This will start the restore process, it start by looking for the `backup.tar.gz` file.
> If it can't find the file in the right path, the script will give out an error.  

Reviewer 2: ```suggestion
This will start the restore process by unpacking the backup file.
If it can't find the file in the provided path, the script will give out an error.  
```
